cameronRomo/depot
# Feature/upgrade cart

## Description

This PR improves on the existing cart functionality. The view for the user is much better, as they can see if they have multiples of a product in their cart, see how much it costs, and see a sum of all the line items in their cart, so they know their grand total before proceeding to checkout. This PR also adds error handling to the cart, so when an invalid cart id is entered in the url, the user is redirected back to the catalog page with an error message. Behind the seens, the error is logged using the Rails logger object.

## Related Issue

closes #5 
closes #6 

## Motivation and Context

The hypothetical client noticed that we were displaying multiples of line items separately, and wanted to indicate multiples with a preceding number and followed by the price of the product. They also noted that entering an invalid cart id in the url caused an error in the application. Concerned about security, they requested that we do something about it.

## How Has This Been Tested?

Controller tests have been updated to ensure that an invalid cart id is not being passed as part of the redirect and tests that a cart is successfully destroyed.

For manual testing:
1. Add products to your cart
2. Navigate to your cart using the URL `http://localhost:3000/carts/1`
3. Confirm that it is rendered as in the screenshot below
4. Click "Empty Cart" and confirm that you are redirected to the catalog page and get a success message at the top indicating your cart is now empty.

## Screenshots (if appropriate):
Current Cart UI:
![Screen Shot 2023-07-18 at 5 22 22 PM](https://github.com/cameronRomo/depot/assets/63012953/62718d9e-2092-44b0-b1e3-1b23607ee98e)

## Types of changes




## Checklist:




### No update to the documentation is needed, but this PR does modify the `line_items` table, and you will need to run a migration.
